File: 1161. Maximum Level Sum of a Binary Tree.java
// Medium
// BFS
// Time:O(n),Space:O(n)
// https://leetcode.cn/problems/maximum-level-sum-of-a-binary-tree/

import java.util.*;
class Solution {
  public int maxLevelSum(TreeNode root) {
      if (root == null) return 0;
      Queue<TreeNode> que = new LinkedList<>();
      que.offer(root);
      int maxSum = Integer.MIN_VALUE, minLevel = 0, level = 1;
      while (!que.isEmpty()) {
          int n = que.size();
          int levelSum = 0;
          for (int i = 0; i < n; i++) {
              TreeNode node = que.poll();
              levelSum += node.val;
              if (node.left != null) que.offer(node.left);
              if (node.right != null) que.offer(node.right);
          }
          if (levelSum > maxSum) {
              maxSum = levelSum;
              minLevel = level;
          }
          level++;
      }
      return minLevel;
  }
}

File: 1001-2000/1161. Maximum Level Sum of a Binary Tree.py
# Time:O(n), Space:O(w)

from collections import deque
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        que = deque([root])
        level = 1
        best_level = 1
        max_sum = float('-inf')

        while que:
            size = len(que)
            cur_sum = 0

            for _ in range(size):
                node = que.popleft()
                cur_sum += node.val

                if node.left:
                    que.append(node.left)
                if node.right:
                    que.append(node.right)
            
            if cur_sum > max_sum:
                max_sum = cur_sum
                best_level = level
            
            level += 1
        
        return best_level
